### ISSUE TITLE
Sony TV IR-remote emulation with push buttons + rotary enc.

### ISSUE THREAD

**Issue by ovi4** [z3t0/Arduino-IRremote#580]
Hi all, Consider the following:
Board Arduino UNO R3, 6 push buttons + IR blaster all on breadboard.

PART 1
The goal: emulate 6 buttons of my sony TV IR remote with an arduino board.
My desired buttons are : Vol+, Vol-, Prog+, Prog-, On/Off and Mute.
So far: I've managed to extract the hex codes for each of the above buttons and wrote the following code.
The problem is with the volume + and -   . It does not seem as responsive and as fast acting as the original IR remote. For instance pressing volume + takes ages to get to maximum volume but using the original IR remote the volume+ gets me to maximum in no time (@ least 3 times faster!!!)
Now I'm thinking, I might need to somehow implement a (repeat code) function but how ?

My level of "C" language is bear minimum I'm an absolute newbie that tries to learn  on the go.
The setup is: pin 3 to IR transmitting LED and pins 2, 4, 5, 6, 7, 8 go to my 6 buttons.
My code (so far) is the following:
``
```
 # include <IRremote.h> // Nice IR Library by Shirriff

 // Define various pins for the push buttons
 const int Vplus = 2;
 const int Vminus = 4;
 const int Pplus = 5;
 const int Pminus = 6;
 const int OnOff = 7;
 const int Mute = 8;

  // Define a variable for each of  the initial button state
 int stateVplus = 0;
 int stateVminus = 0;
 int statePplus = 0;
 int statePminus = 0;
 int stateOnOff = 0;
 int stateMute = 0;

 IRsend irsend;    // Creates  send  Object ???

void setup() {
 // Sets all Push Button pins as INPUTS
  pinMode(Vplus, INPUT);
  pinMode(Vminus, INPUT);  
  pinMode(Pplus, INPUT);
  pinMode(Pminus, INPUT);    
  pinMode(OnOff, INPUT);   
  pinMode(Mute, INPUT); 
}
void loop() {
 // Sets the digital state of each button accordingly
  stateVplus = digitalRead(Vplus);
  stateVminus = digitalRead(Vminus);
  statePplus = digitalRead(Pplus);
  statePminus = digitalRead(Pminus);
  stateOnOff = digitalRead(OnOff);
  stateMute = digitalRead(Mute);

 //If any button is pressed send the appropriate code 
  if (stateVplus == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0x490, 12);   //  Vol+  code
    }
  }
  if (stateVminus == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0xC90, 12);  //  Vol-  code
    }
  }
  if (statePplus == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0x90, 12);  //  Prog+  code
    }
  }
  if (statePminus == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0x890, 12);  //  Prog-  code
    }
  }
  if (stateOnOff == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0xA90, 12);  //  OnOff  code
    }
  }
  if (stateMute == HIGH){
    for (int i = 0; i < 3; i++){
   irsend.sendSony(0x290, 12);  //  Mute  code
    }
  }
      delay(200); // delay between each burst of 3 signals sent (as required by Sony TV to work)
}
```

What I do know is that I have to send every code 3 times per transmission for the TV to respond (according to SONY)  
My question is: how to change my code so that if I keep any button pressed for more than say: 3 seconds the sent code should now become  0xFFFFFFF (a repeat code). I've read some time ago that i have to use a long and name it say : LastCode and any code transmitted for more that 3 seconds should then automatically be store in the above long and then be transmitted as a series of FFFFFFF's instead
but i really have no clue  what to do. Any ideas?

Regarding the wiring of buttons every button has external pull down 10K  resistors ( to ground) 
And finally, after sorting out the issue in my code how do I write my code in a more compact and more elegant manner? It seems too repetitive and long winded :)

PART 2
While tinkering with the code I came up with another idea: What if i substitute the volume and mute buttons with an ordinary mechanical rotary encoder ?  Well I did just that and quickly came up with another code that allows me to use the rotary encoder for vol+ and vol -  but without the mute function (as of now). Here is the code:

```
#include <IRremote.h>

#define outA 2 //CLK 
#define outB 4 //DT
int count = 0;
int precount = 0;
int diffcount = 0;
byte curroutA, curroutB, prevoutA, prevoutB = 0;

IRsend irsend;

void setup() {
  pinMode (outA, INPUT);
  pinMode (outB, INPUT);
  Serial.begin(9600);
  prevoutA = curroutA = digitalRead(outA);
  prevoutB = curroutB = digitalRead(outB);
}

void loop() {
  prevoutA = curroutA;
  prevoutB = curroutB;
  curroutA = digitalRead(outA);
  curroutB = digitalRead(outB);

  if (prevoutA == LOW && prevoutB == LOW && curroutA == HIGH && curroutB == LOW) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      count++; //Clockwise
      irsend.sendSony(0x490, 12);
      Serial.print(count);
      Serial.println("   VOLUME + 0x490");
      delay(20);
    }
  }
  if (prevoutA == HIGH && prevoutB == HIGH && curroutA == LOW && curroutB == HIGH) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      count++; //Clockwise
      irsend.sendSony(0x490, 12);
      Serial.print(count);
      Serial.println("   VOLUME + 0x490");
      delay(20);
    }
  }
  if (prevoutA == LOW && prevoutB == LOW && curroutA == LOW && curroutB == HIGH) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      count--; //Counter-clockwise
      irsend.sendSony(0xC90, 12);
      Serial.print(count);
      Serial.println("   VOLUME - 0xC90");
      delay(20);
    }
  }
  if (prevoutA == HIGH && prevoutB == HIGH && curroutA == HIGH && curroutB == LOW) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      count--; //Counter-clockwise
      irsend.sendSony(0xC90, 12);
      Serial.print(count);
      Serial.println("   VOLUME - 0xC90");
      delay(20);
    }
    delay(100); // between bursts of triple signals
  }
}
```

The code seem to work but....AGAIN!!!!.....very very sluggish slow acting for some unknown reason.
Any hands on help is much appreciated.
Thank you.

**Comment by AnalysIR** [comment]
> What I don't know is how do I tell the loop to run 3 times @ every button press???

It looks like your for loops are doing this already!

```
// Creates IR switch SEND Object (one object per switch)
IRsend sendVplus;
IRsend sendVminus;
IRsend sendPplus;
IRsend sendPminus;
IRsend sendOnOff;
IRsend sendMute;
```
You should be able to use just one object to send every signal!

Might be worth using pullups on the input pins for the switches..

> [And finally, after sorting out the issue in my code how do I write my code in a more compact and more elegant manner? It seems too repetitive, too long winded]

Yes, indeed. I suggest you get it working with one button first. The progress to 2 buttons, using a switch statement to store the hex value in a variable. Then after the switch statement, just dend the sony signal as you already do in one of your for loops. You may need to study up on this first, which is the best way to learn & retain :)

Finally, you didnt place your code inside code tags, which are available in the menu above the edit box <>

Plese re-edit your original post to tidy it up.

**Comment by ovi4** [comment]
Hi, tnx. for your help I've  now edited my post and changed the code and it works to an extend but still issues and,  came up with PART 2. see above .
Kind regards

**Comment by AnalysIR** [comment]
Your attempt at using code tags does not seem to have worked. Try again and remember to use the preview option. It might help to insert the code and select it....then press the <> option for the code tags.
Use preview to veryify that it works.

Also, use the auto-format option in the Arduino IDE to format the code before pasting it in.

Otherwise it is harder to read your code or make sense of it. I didn't try to check your use of an encoder for that reason.

Some issues I noticed:
- The Serial prints inside the for loops will/may mess up the sending of signals.
- the delay(60) for the debounce, will also mess up the signals
- However, you will still need a delay between sending each of the 3 SONY signals in the for loop of approx 20 mSecs

In summary you need to restructure a bit more and consider my earlier suggestion of starting simpler and building it up slowly with more features as you get things working.

**Comment by ovi4** [comment]
Hi, well.... I've had another attempt @ trying to correctly insert my codes in  the comment. Hopefully it worked this time.
On the other hand some days back  I've ordered (online) a book that hopefully will shed some light into my AVR programming skills. The book is called : Make: AVR Programming   Learning to Write Software for Hardware by Elliot Williams.  I hope I've made the right choice since this is gonna be the very first book I've ever purchased regarding programming micro-controllers. 
Thank you once again.

**Comment by AnalysIR** [comment]
The code tags are correct now.

For learning Arduino coding, there are lots of resources on line & probably the best approach is to work thru the various examples supplied with most Arduino libraries.

**Comment by AnalysIR** [comment]
the following may improve things for you (but it is untested)
```
void loop() {
  previousA = currentA;
  previousB = currentB;
  currentA = digitalRead(2);
  currentB = digitalRead(4);

  unsigned long signalValue = 0;
  if (previousA == LOW && previousB == LOW && currentA == HIGH && currentB == LOW) signalValue = 0x490;
  else if (previousA == HIGH && previousB == HIGH && currentA == LOW && currentB == HIGH)  signalValue = 0x490;
  else if (previousA == LOW && previousB == LOW && currentA == LOW && currentB == HIGH)  signalValue = 0xC90;
  else if (previousA == HIGH && previousB == HIGH && currentA == HIGH && currentB == LOW)  signalValue = 0xC90;

  if ( signalValue != 0) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      irsend.sendSony(signalValue, 12);
      delay(20);
    }
    delay(100); // between bursts of triple signals
  }
}
```

The count variable didn't seem to have any purpose, so I removed it.
I also didnt look into the rotary encoder operation and noticed that you didn't set pinMode for pins 2 & 4.

**Comment by ovi4** [comment]
Hi, thank you for the help with your Code. It helped indeed . the speed and responsiveness has now greatly improved. Is closer to the original IR responsiveness. However the range (the distance) is only about 1M or there about. I think that the IR LED that I'm using might not be the right one (is 850nm) i know that certain IR 's have IR  LED's that transmit over 920nm. Also, the 150 ohm resistor that is in series with the IR LED might need to be reduced down to 100 ohms or do I need some sort of 1 transistor amplification ? I really don't know.   Anyways, I have revised the code ( including some change in naming to try shorten some long winded words ) and added the extra (Mute) functionality to the SW button on the Rot. Enc. I've also changed the delay value to 10 milliseconds as response  seems a little smoother . as for the PinMode of 2 and 4 it was my mistake when copy/paste it the code to the comment window. I must have somehow erased it by mistake.
So, here is the code: 

```
#include <IRremote.h> // IR Library by Shirriff

// Define the rotary encoder pins
#define outA 2 //CLK
#define outB 4 //DT
#define sw 8 //SW for the MUTE function

// initial var. state
int stMute, count, precount, diffcount = 0;
byte currA, currB, preA, preB = 0;

IRsend irsend; // define object ?

void setup() {
  pinMode [(sw, outA, outB), INPUT];

  preA = currA = digitalRead(outA);
  preB = currB = digitalRead(outB);
}

void loop() {
  preA = currA;
  preB = currB;
  currA = digitalRead(outA);
  currB = digitalRead(outB);
  stMute = digitalRead(sw);

  unsigned long sigVal = 0;
  if (preA == LOW && preB == LOW && currA == HIGH && currB == LOW) sigVal = 0x490;
  else if (preA == HIGH && preB == HIGH && currA == LOW && currB == HIGH) sigVal = 0x490;
  else if (preA == LOW && preB == LOW && currA == LOW && currB == HIGH) sigVal = 0xC90;
  else if (preA == HIGH && preB == HIGH && currA == HIGH && currB == LOW) sigVal = 0xC90;
  else if (stMute == LOW) sigVal = 0x290;
  if ( sigVal != 0) {
    for (int i = 0; i < 3; i++) {  // sends code 3 times
      irsend.sendSony(sigVal, 12);
      delay(10);
    }
    delay(100); // between bursts of triple signals
  }
}

```

**Comment by AnalysIR** [comment]
I think that the IR LED that I'm using might not be the right one (is 850nm).

Obviously that would not be good. Ususally with 850nm you can see a dim red/pink light. IR for remote control is usually specified @ 940nm or 944nm. TSAL6x00 from Vishay are good quality ones. (where x=1,2 or 4)

100R or even 68R should be OK when directly driven by a standard (5V) Arduino GPIO pin.
_(Putting 2 150R resistors in parallel to get 75R should also be fine)_

> do I need some sort of 1 transistor 

That would help a lot & you can find more information on that by searching.
